### PR TITLE
Add support for dynamic partitioning + fix cache table parsing

### PR DESCRIPTION
**Pull request by robert8138**
### Changes

* Add support so `insert` expression can handle dynamic partitioning (common in Hive & Spark). 
* As part of this, rename partition token used in window function to `partition_by`
* Add exp.Partition 
* Fix parsing logic for `cache` expression for options key value pair. The value should be a string, not a var.
* Add exp.Uncache
* Added tests to verify that parsing works correctly

### Tests
```
............................................
----------------------------------------------------------------------
Ran 44 tests in 0.240s

OK

--------------------------------------------------------------------
Your code has been rated at 10.00/10 (previous run: 10.00/10, +0.00)
```

This is my first time submitting a pull request to sqlglot. PTAL and give me feedback. Thank you!

### CODE REVIEW

File: sqlglot/expressions.py [hunk 1]
```diff
@@ -449,7 +455,7 @@ class TableSample(Expression):
 
 
 class Window(Expression):
-    arg_types = {"this": True, "partition": False, "order": False, "spec": False}
+    arg_types = {"this": True, "partition_by": False, "order": False, "spec": False}
```
Comment (tobymao) [inline sqlglot/expressions.py:458]: why is this changed?

File: sqlglot/generator.py [hunk 2]
```diff
@@ -318,8 +318,21 @@ def insert_sql(self, expression):
         kind = "OVERWRITE" if expression.args.get("overwrite") else "INTO"
         this = self.sql(expression, "this")
         exists = " IF EXISTS " if expression.args.get("exists") else " "
+        partitions = expression.args.get("partition")
+        if partitions:
+            if isinstance(partitions[0], tuple):
+                partition_fields = ", ".join(
```
Comment (tobymao) [inline sqlglot/generator.py:324]: i think i have a helper method, called csv, also can't you just do the check?

csv(f"{x}={y}" if isinstance(partition, tupel) else self.sql(partition) for partition in partitions)

File: sqlglot/generator.py [hunk 3]
```diff
@@ -475,7 +488,7 @@ def where_sql(self, expression):
 
     def window_sql(self, expression):
         this_sql = self.sql(expression, "this")
-        partition = expression.args.get("partition")
+        partition = expression.args.get("partition_by")
```
Comment (tobymao) [inline sqlglot/generator.py:491]: let's not change this for now

File: tests/test_parser.py [hunk 4]
```diff
@@ -74,3 +74,41 @@ def test_function_arguments_validation(self):
 
         with self.assertRaises(ParseError):
             parse_one("IF(a > 0)")
+
+    def test_cache_table(self):
```
Comment (tobymao) [inline tests/test_parser.py:78]: move to test_dialects

File: tests/test_parser.py [hunk 5]
```diff
@@ -74,3 +74,41 @@ def test_function_arguments_validation(self):
...
+        expression = parse_one(
+            "CACHE LAZY TABLE a.b OPTIONS('storageLevel' = 'some_value') AS SELECT x FROM y"
+        )
+        self.assertEqual(expression.args["options"][0].args["this"], "storageLevel")
+        self.assertEqual(expression.args["options"][1].args["this"], "some_value")
+
+    def test_insert_with_dynamic_partitioning(self):
```
Comment (tobymao) [inline tests/test_parser.py:85]: move to identity sql

File: sqlglot/expressions.py [hunk 6]
```diff
@@ -330,7 +330,13 @@ def __hash__(self):
 
 
 class Insert(Expression):
-    arg_types = {"this": True, "expression": True, "overwrite": False, "exists": False}
+    arg_types = {
+        "this": True,
+        "expression": True,
+        "overwrite": False,
+        "exists": False,
+        "partition": False,
```
Comment (tobymao) [inline sqlglot/expressions.py:338]: Make partition first class with 

{"keys": True, "values": False}

Comment (tobymao) [reply sqlglot/expressions.py:338]: or perhaps values true, i'm not sure on waht's cleaner

File: sqlglot/generator.py (hunk 3 above)
Comment (robert8138) [reply sqlglot/generator.py:498]: discussed offline, will keep the distinction here.

File: sqlglot/expressions.py (hunk 6 above)
Comment (robert8138) [reply sqlglot/expressions.py:342]: created exp.Partition with just values. We do not need to track the keys and values separating and do zipping/unzipping.

File: tests/test_parser.py (hunk 5 above)
Comment (robert8138) [reply tests/test_parser.py]: Moved.

File: tests/test_parser.py (hunk 4 above)
Comment (robert8138) [reply tests/test_parser.py]: I put it in identity.sql as well, since I just want to verify parsing works and UNCACHE / CACHE is a Spark concept only.

File: sqlglot/generator.py (hunk 2 above)
Comment (robert8138) [reply sqlglot/generator.py]: Re-implemented this so we don't have to check instancetype

File: sqlglot/expressions.py [hunk 7]
```diff
@@ -330,7 +334,17 @@ def __hash__(self):
...
+    }
+
+
+class Partition(Expression):
+    arg_types = {"values": True}
```
Comment (tobymao) [inline sqlglot/expressions.py:347]: ok, if it's only values, then let's just not override arg_types, just do pass here and use this keyword

File: sqlglot/parser.py [hunk 8]
```diff
@@ -484,6 +495,30 @@ def _parse_cache(self):
             expression=self._parse_cte(),
         )
 
+    def _parse_partition(self):
+        def parse_values():
```
Comment (tobymao) [inline sqlglot/parser.py:499]: we can move this after line 507 so it doesn't parse most of the time

Comment (robert8138) [reply sqlglot/parser.py:505]: Good call. Moved.

**Review by tobymao — APPROVED** [review]
